proxy/http2/Http2ConnectionState.cc

```diff
@@ -1153,6 +1153,13 @@ Http2ConnectionState::state_closed(int event, void *edata)
 Http2Stream *
 Http2ConnectionState::create_stream(Http2StreamId new_id, Http2Error &error)
 {
+  // first check if we've hit the active connection limit
+  if (!ua_session->get_netvc()->add_to_active_queue()) {
```

Comment:
Originally I thought this would be a problem because of multiplexing is HTTP/2 the session may already be on the active queue, but looking at add_to_active_queue, it already deals with that case.

Reply:
Of course in the case of HTTP/2 it may be better if add_to_active queue does not call do_io_close and instead let the error handling in the HTTP/2 logic send a useful error and close the connection.

Reply:
And looking again, that is exactly what the logic does, so this looks good.

Reply:
@shinrich 

Yeah, I tweaked `add_to_active_queue()` to handle the case where the session is already inside the active queue just for the HTTP/2 scenario.

One thing I'm wondering about was whether the throttling should be done with HTTP/2 error `HTTP2_ERROR_REFUSED_STREAM` instead of a generic "internal error". `HTTP2_ERROR_REFUSED_STREAM` will allow the client to retry the request safely. Thoughts?

Reply:
`HTTP2_ERROR_REFUSED_STREAM` (with `HTTP2_ERROR_CLASS_STREAM`) would make more sense. It might cause rapid retries, but in that case the session would be closed with `ENHANCE_YOUR_CALM` eventually because of an excessive error rate.

If we don't want retries, I'd use `NO_ERROR` instead because it's a managed situation but not something unexpected (e.g. memory allocation failure).

Reply:
Yeah, that's exactly my thought as well. If one or few boxes in a cluster are overloaded, a retry from the client is desirable (as it may then end up on a new box). On the other hand, if the entire cluster is overloaded, we don't want to create a retry storm and make matters worse, by allowing a fast retry. I'll change it to `NO_ERROR` as this is intended/designed failure.

Reply:
We cannot use `HTTP2_ERROR_CLASS_NONE`. It's something like null. Valid options to close a connection/stream are `HTTP2_ERROR_CLASS_CONNECTION` or `HTTP2_ERROR_CLASS_STREAM`.

For this case, these two are possible options, IMO.
`HTTP2_ERROR_CLASS_CONNECTION` + `NO_ERROR` : Just close the connection
`HTTP2_ERROR_CLASS_STREAM` + `REFUSED_STREAM` : Close the stream but allows a client to retry on the same connection

`NO_ERROR` does not ensure that clients don't retry. If we just close a connection with `HTTP2_ERROR_CLASS_CONNECTION` + `NO_ERROR`, client may reconnect to a server because the client don't now why the connection was closed. It's basically the same as HTTP/1.1 behavior.

Reasons I suggested `REFUSED_STREAM` are 1) it's relatively lighter than retry on a new connection, and 2) it would end up with `ENHANCE_YOUR_CALM`, which suggests the behavior is generating excessive load, if retries seem like making things worse. There is still no guarantee but I hope the error code ceases retries.

Another option is retuning 503, (and then `HTTP2_ERROR_CLASS_CONNECTION` + `NO_ERROR`). Most clients would not retry soon if they received it.

Reply:
That’s not correct. Most clients will retry as soon as they see a 503 including automatically at the Network layer. Infact, 502/503 are considered the safest status codes for the clients to automatically retry.

I’ll just change it back to return (connection class + no error). I don’t think this is a big deal and the only way to prevent a retry would be If there was a custom protocol between server and client.

Reply:
It's a bummer if clients ignore Retry-After header sent with 503.